lib/web_ui/lib/src/engine/canvaskit/text_fragmenter.dart

```diff
@@ -0,0 +1,91 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:typed_data';
+
+import '../dom.dart';
+import '../text/line_breaker.dart';
+
+abstract class CkTextFragmenter {
+  const CkTextFragmenter(this.text);
+
+  /// The text to be fragmented.
+  final String text;
+
+  /// Performs the fragmenting of [text] and returns a [Uint32List].
+  Uint32List fragment();
+}
+
+/// The granularity at which to segment text.
+///
+/// To find all supported granularities, see:
+/// - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter/Segmenter
+enum IntlSegmenterGranularity {
+  grapheme,
+  word,
+}
+
+class CkIntlFragmenter extends CkTextFragmenter {
+  CkIntlFragmenter(super.text, this.granularity)
+      : assert(domIntl.Segmenter != null);
+
+  final IntlSegmenterGranularity granularity;
+
+  String get _granularityString {
+    switch (granularity) {
+      case IntlSegmenterGranularity.grapheme:
+        return 'grapheme';
+      case IntlSegmenterGranularity.word:
+        return 'word';
+    }
+  }
+
+  @override
+  Uint32List fragment() {
+    final DomIteratorWrapper<DomSegment> iterator =
+        createIntlSegmenter(granularity: _granularityString)
+            .segment(text)
+            .iterator();
+
+    final List<int> breaks = <int>[];
+    while (iterator.moveNext()) {
+      breaks.add(iterator.current.index);
+    }
+    breaks.add(text.length);
+
+    return Uint32List.fromList(breaks);
+  }
+}
+
+// These are the soft/hard line break values expected by SkParagraph.
+const int _kSoftLineBreak = 0;
+const int _kHardLineBreak = 1;
+
+class CkLineBreakFragmenter extends CkTextFragmenter {
+  CkLineBreakFragmenter(super.text) : _fragmenter = LineBreakFragmenter(text);
+
+  final LineBreakFragmenter _fragmenter;
+
+  @override
+  Uint32List fragment() {
+    final List<LineBreakFragment> fragments = _fragmenter.fragment();
+
+    final int size = (fragments.length + 1) * 2;
+    final Uint32List typedArray = Uint32List(size);
```

Comment:
Consider using `CanvasKit.Malloc` to avoid creating a temporary array and extra copying into wasm.

Reply:
Up until today, I didn't really understand the benefit of using CK's Malloc. It seemed to me like the CK bindings always end up copying the arrays anyway.

But I took a closer look now, and I think I understand where this optimization is happening on the CK side:
https://github.com/google/skia/blob/deb1cf8d127986fee464946a12b248d6f17fb629/modules/canvaskit/memory.js#L126-L128

What threw me off was the name `copy1dArray`. But it turns out this function doesn't always *copy*. It first checks if the array was malloc'd and avoids the copying in that case.